plugins/modem-manager/fu-sahara-loader.c

```diff
@@ -0,0 +1,559 @@
+		fu_sahara_loader_send_packet(self, ping, NULL);
+		rx_packet = fu_sahara_loader_qdl_read(self, error);
+	}
+
+	g_return_val_if_fail(rx_packet != NULL, FALSE);
+
+	if (g_getenv("FWUPD_MODEM_MANAGER_VERBOSE") != NULL)
+		fu_common_dump_raw(G_LOG_DOMAIN, "rx packet", rx_packet->data, rx_packet->len);
+
+	if (sahara_packet_get_command_id(rx_packet) != SAHARA_HELLO_ID) {
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_FAILED,
+			    "Received a different packet while waiting for the HELLO packet");
+		fu_sahara_loader_send_reset_packet(self, NULL);
+		return FALSE;
+	}
+
+	tx_packet = fu_sahara_loader_compose_hello_response_packet(SAHARA_MODE_IMAGE_TX_PENDING);
+
+	return fu_sahara_loader_send_packet(self, tx_packet, error);
+}
+
+/* main routine */
+gboolean
+fu_sahara_loader_run(FuSaharaLoader *self, GBytes *prog, GError **error)
+{
+	g_return_val_if_fail(prog != NULL, FALSE);
+
+	if (g_getenv("FWUPD_MODEM_MANAGER_VERBOSE") != NULL)
+		g_debug("STATE -- SAHARA_WAIT_HELLO");
+	if (!fu_sahara_loader_wait_hello_rsp(self, error))
+		return FALSE;
+
+	while (TRUE) {
+		guint32 command_id;
+		struct sahara_packet *pkt;
+		g_autoptr(GByteArray) rx_packet = NULL;
+		g_autoptr(GByteArray) tx_packet = NULL;
+		g_autoptr(GError) error_local = NULL;
+
+		if (g_getenv("FWUPD_MODEM_MANAGER_VERBOSE") != NULL)
+			g_debug("STATE -- SAHARA_WAIT_COMMAND");
+		rx_packet = fu_sahara_loader_qdl_read(self, error);
+		if (rx_packet == NULL)
+			break;
+		if (rx_packet->len != sahara_packet_get_length(rx_packet)) {
+			g_set_error(error,
+				    G_IO_ERROR,
+				    G_IO_ERROR_FAILED,
+				    "Received packet length is not matching");
+			break;
+		}
+
+		if (g_getenv("FWUPD_MODEM_MANAGER_VERBOSE") != NULL) {
+			fu_common_dump_raw(G_LOG_DOMAIN,
+					   "rx_packet",
+					   rx_packet->data,
+					   rx_packet->len);
+		}
+
+		command_id = sahara_packet_get_command_id(rx_packet);
+		pkt = (struct sahara_packet *)(rx_packet->data);
+		if (command_id == SAHARA_HELLO_ID) {
+			tx_packet = fu_sahara_loader_compose_hello_response_packet(
+			    SAHARA_MODE_IMAGE_TX_PENDING);
+			fu_sahara_loader_send_packet(self, tx_packet, &error_local);
+		} else if (command_id == SAHARA_READ_DATA_ID) {
+			guint32 offset = pkt->read_data.offset;
+			guint32 length = pkt->read_data.length;
+			fu_sahara_loader_write_prog(self, offset, length, prog, &error_local);
+		} else if (command_id == SAHARA_READ_DATA_64_BIT_ID) {
+			guint64 offset = pkt->read_data_64bit.offset;
+			guint64 length = pkt->read_data_64bit.length;
+			fu_sahara_loader_write_prog(self, offset, length, prog, &error_local);
+		} else if (command_id == SAHARA_END_OF_IMAGE_TX_ID) {
+			guint32 status = pkt->end_of_image_transfer.status;
+			if (status == SAHARA_STATUS_SUCCESS) {
+				tx_packet = fu_sahara_loader_compose_done_packet();
+				fu_sahara_loader_send_packet(self, tx_packet, &error_local);
+			}
+		} else if (command_id == SAHARA_DONE_RESP_ID) {
+			return TRUE;
+		} else {
```

Comment:
you don;t need else if you returned

Reply:
I feel like i still need else here: I only need to print this warning if the previous `if` statements didn't match, and I only return in one of them

Reply:
Yup, fair enough. No worries.